File: src/main/java/de/pflugradts/passbird/application/commandhandling/handler/namespace/CanListAvailableNamespaces.java
package de.pflugradts.passbird.application.commandhandling.handler.namespace;

import de.pflugradts.passbird.domain.model.namespace.Namespace;
import de.pflugradts.passbird.domain.service.FixedNamespaceService;

import java.util.Optional;
import java.util.stream.Collectors;

public interface CanListAvailableNamespaces {

    default boolean hasCustomNamespaces(final FixedNamespaceService namespaceService) {
        return namespaceService.all().anyMatch(Optional::isPresent);
    }

    default String getAvailableNamespaces(final FixedNamespaceService namespaceService, final boolean includeCurrent) {
        final var defaultNamespaceLine = includeCurrent || !namespaceService.getCurrentNamespace().equals(Namespace.Companion.getDEFAULT())
            ? "\t0: " + Namespace.Companion.getDEFAULT().getBytes().asString() + System.lineSeparator() : "";
        return defaultNamespaceLine + namespaceService.all()
            .filter(Optional::isPresent)
            .filter(namespace -> includeCurrent || !namespace.get().equals(namespaceService.getCurrentNamespace()))
            .map(Optional::get)
            .map(namespace -> "\t"
                + namespace.getSlot().index()
                + ": "
                + namespace.getBytes().asString()
                + System.lineSeparator())
            .collect(Collectors.joining());
    }

}

File: src/main/kotlin/de/pflugradts/passbird/application/commandhandling/handler/namespace/CanListAvailableNamespaces.kt
package de.pflugradts.passbird.application.commandhandling.handler.namespace

import de.pflugradts.passbird.domain.model.namespace.Namespace.Companion.DEFAULT
import de.pflugradts.passbird.domain.service.NamespaceService

abstract class CanListAvailableNamespaces(
    private val namespaceService: NamespaceService,
) {
    fun hasCustomNamespaces() = namespaceService.all().anyMatch { it.isPresent }
    fun getAvailableNamespaces(includeCurrent: Boolean) =
        (if (includeCurrent || namespaceService.getCurrentNamespace() != DEFAULT) "\t0: ${DEFAULT.bytes.asString()}\n" else "") +
            namespaceService.all()
                .filter { it.isPresent }
                .map { it.get() }
                .filter { includeCurrent || it != namespaceService.getCurrentNamespace() }
                .map { "\t${it.slot.index()}: ${it.bytes.asString()}" }
                .toList().joinToString("\n")
}
